harness/determined/profiler.py

```diff
@@ -579,21 +585,23 @@ def run(self) -> None:
 class ProfilerSenderThread(threading.Thread):
     """
     This is a thread that exists solely so that we can make API calls without blocking.
-    It has a Queue through which work is sent to the thread
+    It has a Queue through which work is sent to the thread. It is aware of the number of
+    upstream producers and exits whenever it receives a ShutdownMessage from each producer.
     """
 
-    def __init__(self, inbound_queue: queue.Queue, master_url: str) -> None:
+    def __init__(self, inbound_queue: queue.Queue, master_url: str, num_producers: int) -> None:
         self.master_url = master_url
         self.inbound_queue = inbound_queue
+        self.num_producers = num_producers
+        self.producers_shutdown = 0
         super().__init__(daemon=True)
 
-    def send_shutdown_signal(self) -> None:
-        self.inbound_queue.put(ShutdownMessage())
-
     def run(self) -> None:
         while True:
             message = self.inbound_queue.get()
             if isinstance(message, ShutdownMessage):
+                self.producers_shutdown += 1
+            if self.num_producers == self.producers_shutdown:
```

Comment:
nit: I would indent this check so it only happens when a `ShutdownMessage` is received. Same outcome, but a little easier to follow IMO. Up to you.